File: src/main/java/com/uiptv/db/DatabasePatchesUtils.java
package com.uiptv.db;

import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.security.MessageDigest;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.time.Instant;
import java.util.ArrayList;
import java.util.HexFormat;
import java.util.List;

public class DatabasePatchesUtils {
    private static final String MIGRATIONS_LIST_RESOURCE = "db/migrations/migrations.txt";
    private static final String BASELINE_RESOURCE = "db/migrations/0000_baseline.sql";
    private static final String MIGRATIONS_DIR_RESOURCE = "db/migrations/";

    public static void applyPatches(Connection conn) throws SQLException {
        createSchemaMigrationsTable(conn);
        List<String> migrationNames = readMigrationNames();
        for (String migrationName : migrationNames) {
            applyMigration(conn, migrationName);
        }
    }

    public static void applyBaseline(Connection conn) throws SQLException {
        String baselineSql = readResource(BASELINE_RESOURCE);
        executeMigrationContent(conn, baselineSql);
    }

    public static boolean hasMigrationsListResource() {
        return resourceExists(MIGRATIONS_LIST_RESOURCE);
    }

    public static boolean hasBaselineResource() {
        return resourceExists(BASELINE_RESOURCE);
    }

    private static void applyMigration(Connection conn, String migrationName) throws SQLException {
        String resourcePath = MIGRATIONS_DIR_RESOURCE + migrationName;
        String migrationSql = readResource(resourcePath);
        String checksum = checksum(migrationSql);

        MigrationRecord existing = findMigrationRecord(conn, migrationName);
        if (existing != null && "success".equalsIgnoreCase(existing.status) && checksum.equals(existing.checksum)) {
            return;
        }

        boolean originalAutoCommit = conn.getAutoCommit();
        try {
            conn.setAutoCommit(false);
            executeMigrationContent(conn, migrationSql);
            upsertMigrationRecord(conn, migrationName, checksum, "success", null);
            conn.commit();
        } catch (Exception ex) {
            conn.rollback();
            upsertMigrationRecord(conn, migrationName, checksum, "failed", safeError(ex));
            conn.commit();
        } finally {
            conn.setAutoCommit(originalAutoCommit);
        }
    }

    private static void executeMigrationContent(Connection conn, String migrationSql) throws SQLException {
        String directive = findDirectiveLine(migrationSql);
        if (directive != null) {
            executeDirective(conn, directive);
            return;
        }
        executeSqlStatements(conn, migrationSql);
    }

    private static void executeDirective(Connection conn, String directiveLine) throws SQLException {
        String[] parts = directiveLine.trim().split("\\s+", 4);
        if (parts.length < 3) {
            throw new SQLException("Invalid migration directive: " + directiveLine);
        }
        String command = parts[0].toLowerCase();
        if ("--@add_column".equals(command)) {
            if (parts.length < 4) {
                throw new SQLException("Invalid add_column directive: " + directiveLine);
            }
            String table = parts[1];
            String column = parts[2];
            String definition = parts[3];
            if (columnExists(conn, table, column)) {
                return;
            }
            try (Statement stmt = conn.createStatement()) {
                stmt.executeUpdate("ALTER TABLE " + table + " ADD COLUMN " + column + " " + definition);
            }
            return;
        }
        if ("--@drop_column".equals(command)) {
            String table = parts[1];
            String column = parts[2];
            if (!columnExists(conn, table, column)) {
                return;
            }
            try (Statement stmt = conn.createStatement()) {
                stmt.executeUpdate("ALTER TABLE " + table + " DROP COLUMN " + column);
            }
            return;
        }
        throw new SQLException("Unsupported migration directive: " + directiveLine);
    }

    private static void executeSqlStatements(Connection conn, String migrationSql) throws SQLException {
        StringBuilder builder = new StringBuilder();
        String[] lines = migrationSql.split("\\R");
        for (String line : lines) {
            String trimmed = line.trim();
            if (trimmed.startsWith("--")) {
                continue;
            }
            builder.append(line).append('\n');
        }
        String sqlBlob = builder.toString();
        for (String raw : sqlBlob.split(";")) {
            String statement = raw.trim();
            if (statement.isEmpty()) {
                continue;
            }
            try (Statement stmt = conn.createStatement()) {
                stmt.executeUpdate(statement);
            }
        }
    }

    private static boolean columnExists(Connection conn, String tableName, String columnName) throws SQLException {
        String sql = "PRAGMA table_info(" + tableName + ")";
        try (Statement stmt = conn.createStatement(); ResultSet rs = stmt.executeQuery(sql)) {
            while (rs.next()) {
                if (columnName.equalsIgnoreCase(rs.getString("name"))) {
                    return true;
                }
            }
            return false;
        }
    }

    private static String findDirectiveLine(String migrationSql) {
        String[] lines = migrationSql.split("\\R");
        for (String line : lines) {
            String trimmed = line.trim();
            if (trimmed.startsWith("--@")) {
                return trimmed;
            }
        }
        return null;
    }

    private static List<String> readMigrationNames() {
        String content = readResource(MIGRATIONS_LIST_RESOURCE);
        List<String> names = new ArrayList<>();
        String[] lines = content.split("\\R");
        for (String line : lines) {
            String trimmed = line.trim();
            if (trimmed.isEmpty() || trimmed.startsWith("#")) {
                continue;
            }
            names.add(trimmed);
        }
        return names;
    }

    private static String readResource(String path) {
        try (InputStream in = openResource(path)) {
            if (in == null) {
                throw new IllegalStateException("Migration resource not found: " + path);
            }
            StringBuilder out = new StringBuilder();
            try (BufferedReader reader = new BufferedReader(new InputStreamReader(in, StandardCharsets.UTF_8))) {
                String line;
                while ((line = reader.readLine()) != null) {
                    out.append(line).append('\n');
                }
            }
            return out.toString();
        } catch (Exception ex) {
            throw new IllegalStateException("Unable to read migration resource: " + path, ex);
        }
    }

    private static boolean resourceExists(String path) {
        try (InputStream in = openResource(path)) {
            return in != null;
        } catch (Exception ex) {
            return false;
        }
    }

    private static InputStream openResource(String path) {
        String normalized = path.startsWith("/") ? path.substring(1) : path;
        String absolute = "/" + normalized;

        InputStream in = DatabasePatchesUtils.class.getResourceAsStream(absolute);
        if (in != null) {
            return in;
        }

        ClassLoader contextClassLoader = Thread.currentThread().getContextClassLoader();
        if (contextClassLoader != null) {
            in = contextClassLoader.getResourceAsStream(normalized);
            if (in != null) {
                return in;
            }
        }

        ClassLoader classLoader = DatabasePatchesUtils.class.getClassLoader();
        if (classLoader != null) {
            in = classLoader.getResourceAsStream(normalized);
            if (in != null) {
                return in;
            }
        }

        Module module = DatabasePatchesUtils.class.getModule();
        try {
            in = module.getResourceAsStream(normalized);
            if (in != null) {
                return in;
            }
        } catch (Exception ignored) {
            // Continue trying other strategies.
        }

        Path localPath = Paths.get("src", "main", "resources", normalized);
        if (Files.exists(localPath)) {
            try {
                return new FileInputStream(localPath.toFile());
            } catch (Exception ignored) {
                return null;
            }
        }
        return null;
    }

    private static String checksum(String text) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hash = digest.digest(text.getBytes(StandardCharsets.UTF_8));
            return HexFormat.of().formatHex(hash);
        } catch (Exception ex) {
            throw new IllegalStateException("Unable to compute migration checksum", ex);
        }
    }

    private static void createSchemaMigrationsTable(Connection conn) throws SQLException {
        try (Statement stmt = conn.createStatement()) {
            stmt.executeUpdate("CREATE TABLE IF NOT EXISTS schema_migrations ("
                    + "name TEXT PRIMARY KEY,"
                    + "checksum TEXT NOT NULL,"
                    + "status TEXT NOT NULL,"
                    + "applied_at INTEGER NOT NULL,"
                    + "error_message TEXT"
                    + ")");
        }
    }

    private static MigrationRecord findMigrationRecord(Connection conn, String name) throws SQLException {
        String sql = "SELECT checksum, status FROM schema_migrations WHERE name = ?";
        try (PreparedStatement ps = conn.prepareStatement(sql)) {
            ps.setString(1, name);
            try (ResultSet rs = ps.executeQuery()) {
                if (!rs.next()) {
                    return null;
                }
                return new MigrationRecord(rs.getString("checksum"), rs.getString("status"));
            }
        }
    }

    private static void upsertMigrationRecord(Connection conn, String name, String checksum, String status, String errorMessage) throws SQLException {
        String sql = "INSERT INTO schema_migrations(name, checksum, status, applied_at, error_message) VALUES(?,?,?,?,?) "
                + "ON CONFLICT(name) DO UPDATE SET "
                + "checksum=excluded.checksum,"
                + "status=excluded.status,"
                + "applied_at=excluded.applied_at,"
                + "error_message=excluded.error_message";
        try (PreparedStatement ps = conn.prepareStatement(sql)) {
            ps.setString(1, name);
            ps.setString(2, checksum);
            ps.setString(3, status);
            ps.setLong(4, Instant.now().getEpochSecond());
            ps.setString(5, errorMessage);
            ps.executeUpdate();
        }
    }

    private static String safeError(Exception ex) {
        String message = ex == null ? "" : ex.getMessage();
        if (message == null || message.isBlank()) {
            return ex == null ? "" : ex.getClass().getSimpleName();
        }
        if (message.length() > 1000) {
            return message.substring(0, 1000);
        }
        return message;
    }

    private record MigrationRecord(String checksum, String status) {
    }
}

File: server/src/main/kotlin/com/uiptv/db/DatabasePatchesUtils.kt
package com.uiptv.db

import java.io.BufferedReader
import java.io.FileInputStream
import java.io.InputStream
import java.io.InputStreamReader
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.security.MessageDigest
import java.sql.Connection
import java.sql.PreparedStatement
import java.sql.ResultSet
import java.sql.SQLException
import java.sql.Statement
import java.time.Instant
import java.util.HexFormat

class DatabasePatchesUtils private constructor() {
    companion object {
        private const val MIGRATIONS_LIST_RESOURCE = "db/migrations/migrations.txt"
        private const val BASELINE_RESOURCE = "db/migrations/0000_baseline.sql"
        private const val MIGRATIONS_DIR_RESOURCE = "db/migrations/"

        @JvmStatic
        @Throws(SQLException::class)
        fun applyPatches(conn: Connection) {
            createSchemaMigrationsTable(conn)
            for (migrationName in readMigrationNames(MIGRATIONS_LIST_RESOURCE)) {
                applyMigration(conn, migrationName)
            }
        }

        @JvmStatic
        @Throws(SQLException::class)
        fun applyBaseline(conn: Connection) {
            executeMigrationContent(conn, readResource(BASELINE_RESOURCE))
        }

        @JvmStatic
        fun hasMigrationsListResource(): Boolean = resourceExists(MIGRATIONS_LIST_RESOURCE)

        @JvmStatic
        fun hasBaselineResource(): Boolean = resourceExists(BASELINE_RESOURCE)

        @Throws(SQLException::class)
        private fun applyMigration(conn: Connection, migrationName: String) {
            val resourcePath = MIGRATIONS_DIR_RESOURCE + migrationName
            val migrationSql = readResource(resourcePath)
            val checksum = checksum(migrationSql)
            val existing = findMigrationRecord(conn, migrationName)
            if (existing != null && existing.status.equals("success", true) && checksum == existing.checksum) {
                return
            }

            val originalAutoCommit = conn.autoCommit
            try {
                conn.autoCommit = false
                executeMigrationContent(conn, migrationSql)
                upsertMigrationRecord(conn, migrationName, checksum, "success", null)
                conn.commit()
            } catch (ex: Exception) {
                conn.rollback()
                upsertMigrationRecord(conn, migrationName, checksum, "failed", safeError(ex))
                conn.commit()
            } finally {
                conn.autoCommit = originalAutoCommit
            }
        }

        @Throws(SQLException::class)
        @JvmStatic
        private fun executeMigrationContent(conn: Connection, migrationSql: String) {
            val directive = findDirectiveLine(migrationSql)
            if (directive != null) {
                executeDirective(conn, directive)
            } else {
                executeSqlStatements(conn, migrationSql)
            }
        }

        @Throws(SQLException::class)
        @JvmStatic
        private fun executeDirective(conn: Connection, directiveLine: String) {
            val parts = directiveLine.trim().split(Regex("\\s+"), limit = 4)
            if (parts.size < 3) {
                throw SQLException("Invalid migration directive: $directiveLine")
            }
            val command = parts[0].lowercase()
            if (command == "--@add_column") {
                if (parts.size < 4) {
                    throw SQLException("Invalid add_column directive: $directiveLine")
                }
                val table = parts[1]
                val column = parts[2]
                val definition = parts[3]
                if (columnExists(conn, table, column)) {
                    return
                }
                conn.createStatement().use { stmt ->
                    stmt.executeUpdate("ALTER TABLE $table ADD COLUMN $column $definition")
                }
                return
            }
            if (command == "--@drop_column") {
                val table = parts[1]
                val column = parts[2]
                if (!columnExists(conn, table, column)) {
                    return
                }
                conn.createStatement().use { stmt ->
                    stmt.executeUpdate("ALTER TABLE $table DROP COLUMN $column")
                }
                return
            }
            throw SQLException("Unsupported migration directive: $directiveLine")
        }

        @Throws(SQLException::class)
        @JvmStatic
        private fun executeSqlStatements(conn: Connection, migrationSql: String) {
            val sqlBlob = buildString {
                migrationSql.split(Regex("\\R")).forEach { line ->
                    if (!line.trim().startsWith("--")) {
                        append(line).append('\n')
                    }
                }
            }
            sqlBlob.split(";").map(String::trim).filter(String::isNotEmpty).forEach { statement ->
                conn.createStatement().use { stmt -> stmt.executeUpdate(statement) }
            }
        }

        @Throws(SQLException::class)
        @JvmStatic
        private fun columnExists(conn: Connection, tableName: String, columnName: String): Boolean {
            conn.createStatement().use { stmt ->
                stmt.executeQuery("PRAGMA table_info($tableName)").use { rs ->
                    while (rs.next()) {
                        if (columnName.equals(rs.getString("name"), true)) {
                            return true
                        }
                    }
                }
            }
            return false
        }

        @JvmStatic
        private fun findDirectiveLine(migrationSql: String): String? =
            migrationSql.split(Regex("\\R"))
                .map(String::trim)
                .firstOrNull { it.startsWith("--@") }

        @JvmStatic
        private fun readMigrationNames(resourcePath: String): List<String> =
            readResource(resourcePath)
                .split(Regex("\\R"))
                .map(String::trim)
                .filter { it.isNotEmpty() && !it.startsWith("#") }

        @JvmStatic
        private fun readResource(path: String): String {
            try {
                openResource(path)?.use { input ->
                    BufferedReader(InputStreamReader(input, StandardCharsets.UTF_8)).use { reader ->
                        return buildString {
                            while (true) {
                                val line = reader.readLine() ?: break
                                append(line).append('\n')
                            }
                        }
                    }
                }
            } catch (e: Exception) {
                throw IllegalStateException("Unable to read migration resource: $path", e)
            }
            throw IllegalStateException("Migration resource not found: $path")
        }

        @JvmStatic
        private fun resourceExists(path: String): Boolean =
            try {
                openResource(path)?.use { true } ?: false
            } catch (_: Exception) {
                false
            }

        @JvmStatic
        private fun openResource(path: String): InputStream? {
            val normalized = if (path.startsWith("/")) path.substring(1) else path
            val absolute = "/$normalized"

            DatabasePatchesUtils::class.java.getResourceAsStream(absolute)?.let { return it }
            Thread.currentThread().contextClassLoader?.getResourceAsStream(normalized)?.let { return it }
            DatabasePatchesUtils::class.java.classLoader?.getResourceAsStream(normalized)?.let { return it }
            try {
                DatabasePatchesUtils::class.java.module.getResourceAsStream(normalized)?.let { return it }
            } catch (_: Exception) {
            }
            val localPath: Path = Paths.get("src", "main", "resources", normalized)
            return if (Files.isRegularFile(localPath)) FileInputStream(localPath.toFile()) else null
        }

        @JvmStatic
        private fun checksum(text: String): String =
            try {
                val digest = MessageDigest.getInstance("SHA-256")
                HexFormat.of().formatHex(digest.digest(text.toByteArray(StandardCharsets.UTF_8)))
            } catch (ex: Exception) {
                throw IllegalStateException("Unable to compute migration checksum", ex)
            }

        @Throws(SQLException::class)
        @JvmStatic
        private fun createSchemaMigrationsTable(conn: Connection) {
            conn.createStatement().use { stmt ->
                stmt.executeUpdate(
                    "CREATE TABLE IF NOT EXISTS schema_migrations (" +
                        "name TEXT PRIMARY KEY," +
                        "checksum TEXT NOT NULL," +
                        "status TEXT NOT NULL," +
                        "applied_at INTEGER NOT NULL," +
                        "error_message TEXT" +
                        ")"
                )
            }
        }

        @Throws(SQLException::class)
        @JvmStatic
        private fun findMigrationRecord(conn: Connection, name: String): MigrationRecord? {
            conn.prepareStatement("SELECT checksum, status FROM schema_migrations WHERE name = ?").use { ps ->
                ps.setString(1, name)
                ps.executeQuery().use { rs ->
                    if (!rs.next()) {
                        return null
                    }
                    return MigrationRecord(rs.getString("checksum"), rs.getString("status"))
                }
            }
        }

        @Throws(SQLException::class)
        @JvmStatic
        private fun upsertMigrationRecord(conn: Connection, name: String, checksum: String, status: String, errorMessage: String?) {
            val sql = "INSERT INTO schema_migrations(name, checksum, status, applied_at, error_message) VALUES(?,?,?,?,?) " +
                "ON CONFLICT(name) DO UPDATE SET " +
                "checksum=excluded.checksum," +
                "status=excluded.status," +
                "applied_at=excluded.applied_at," +
                "error_message=excluded.error_message"
            conn.prepareStatement(sql).use { ps ->
                ps.setString(1, name)
                ps.setString(2, checksum)
                ps.setString(3, status)
                ps.setLong(4, Instant.now().epochSecond)
                ps.setString(5, errorMessage)
                ps.executeUpdate()
            }
        }

        @JvmStatic
        private fun safeError(ex: Exception?): String {
            val message = ex?.message.orEmpty()
            if (message.isBlank()) {
                return ex?.javaClass?.simpleName.orEmpty()
            }
            return if (message.length > 1000) message.substring(0, 1000) else message
        }
    }

    private data class MigrationRecord(val checksum: String, val status: String)
}
